### ISSUE TITLE
why rewards in "multi_agent_different_spaces_for_agents" with "MADDPG" remain nan. ?

### ISSUE THREAD

**Issue by SaraRezaei** [ray-project/ray#28951]
I faced this issue and I can't solve it. Do any one have a solution for it? thank you in advance.

from __future__ import absolute_import
import argparse
import gym
from gym.spaces import Dict, Discrete, MultiDiscrete, Tuple, Box
import numpy as np
import os
import random
import ray
from ray import air, tune
from ray.rllib.algorithms import ppo
from ray.rllib.env.env_context import EnvContext
from ray.rllib.models import ModelCatalog
from ray.rllib.models.tf.tf_modelv2 import TFModelV2
from ray.rllib.models.tf.fcnet import FullyConnectedNetwork
from ray.rllib.models.torch.torch_modelv2 import TorchModelV2
from ray.rllib.models.torch.fcnet import FullyConnectedNetwork as TorchFC
from ray.rllib.utils.framework import try_import_tf, try_import_torch
from ray.rllib.utils.test_utils import check_learning_achieved
from ray.tune.logger import pretty_print
#-------------------------------------------------------------------------------------
from ray.rllib.env.multi_agent_env import MultiAgentEnv, ENV_STATE
from ray.rllib.policy.policy import PolicySpec 
import numpy as np  
#=====================================================================================
tf1, tf, tfv = try_import_tf()
torch, nn = try_import_torch()

parser = argparse.ArgumentParser()
parser.add_argument(
    "--run", type=str, default="MADDPG", help="The RLlib-registered algorithm to use." #PPO
)
parser.add_argument(
    "--framework",
    choices=["tf", "tf2", "tfe", "torch"],
    default="tf",
    help="The DL framework specifier.",
)
parser.add_argument(
    "--as-test",
    action="store_true",
    help="Whether this script should be run as a test: --stop-reward must "
    "be achieved within --stop-timesteps AND --stop-iters.",
)
parser.add_argument(
    "--stop-iters", type=int, default=40, help="Number of iterations to train."
)
parser.add_argument(
    "--stop-timesteps", type=int, default=100000, help="Number of timesteps to train."
)
parser.add_argument(
    "--stop-reward", type=float, default=1.0, help="Reward at which we stop training."
)
parser.add_argument(
    "--no-tune",
    action="store_true",
    help="Run without Tune using a manual train loop instead. In this case,"
    "use PPO without grid search and no TensorBoard.",
)
parser.add_argument(
    "--local-mode",
    action="store_true",
    help="Init Ray in local mode for easier debugging.",
)
#=====================================================================================
class My_env(MultiAgentEnv):
    def __init__(self, env_config):
        self.agents = {0, 1}
        self._agent_ids = set(self.agents)

        self.dones = set()

        self._spaces_in_preferred_format = True
        self.observation_space = gym.spaces.Dict(
            {
                0: gym.spaces.Box(low=-1.0, high=1.0, shape=(10,)),
                1: gym.spaces.Box(low=-1.0, high=1.0, shape=(20,)),
            }
        )
        self.action_space = gym.spaces.Dict(
            {0: gym.spaces.Discrete(2), 1: gym.spaces.Discrete(3)}
        )

        super().__init__()
    #----------------------------------------------------------------------------------
    def reset(self):
        self.dones = set()
        self.step_count = 0
        return {i: self.observation_space[i].sample() for i in self.agents}
    #----------------------------------------------------------------------------------
    def step(self, action_dict):
        obs, rew, done, info = {}, {}, {}, {}
        for i, action in action_dict.items():
            obs[i] = self.observation_space[i].sample()
            rew[i] = 0.0
            done[i] = False
            info[i] = {}
        done["__all__"] = len(self.dones) == len(self.agents)
        return obs, rew, done, info
#=====================================================================================
if __name__ == "__main__":
    args = parser.parse_args()

    ray.init(local_mode=args.local_mode)
    #----------------------------------------------------------------------------------
    config = {
        "env": My_env,
        "env_config": {
            "actions_are_logits": True,
        },
        "num_workers": 1,  # parallelism
        #"num_steps_sampled_before_learning_starts": 100,
        "multiagent": {
            "policies": {
                "pol1": PolicySpec(
                    config={"agent_id": 0},
                ),
                "pol2": PolicySpec(
                    config={"agent_id": 1},
                ),
            },
            "policy_mapping_fn": (lambda aid, episode, worker, **kwargs: "pol2" if aid else "pol1"),
        },
        "framework": args.framework,
        "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
    }  #=====================================================================================
    stop = {
        "episode_reward_mean": args.stop_reward,
        "timesteps_total": args.stop_timesteps,
        "training_iteration": args.stop_iters,
    }
  #===========================================================================================
    results = tune.Tuner(
        args.run,
        run_config=air.RunConfig(stop=stop, verbose=2),
        param_space=config,
    ).fit()
   #=====================================================================================
    if args.as_test:
        check_learning_achieved(results, args.stop_reward)
    ray.shutdown()

**Comment by mvindiola1** [comment]
Hi @SaraRezaei,

From the code sample you shared your environment never ends. done is always False in step. This is why. You can try setting done to True after some number of steps or set the config option "horizon" to some positive number to have rllib terminate the episode after that many steps.

**Comment by SaraRezaei** [comment]
Hi @mvindiola1 
that's right. 
thank you.
